ee/clickhouse/client.py

```diff
@@ -166,6 +165,23 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
                     save_query(prepared_sql, execution_time)
         return result
 
+    def substitute_params(query, params):
+        """
+        Helper method to ease rendering of sql clickhouse queries progressively.
+        For example, there are many places where we construct queries to be used
+        as subqueries of others. Each time we generate a subquery we also pass
+        up the "bound" parameters to be used to render the query, which
+        otherwise only happens at the point of calling clickhouse via the
+        clickhouse_driver `Client`.
+
+        This results in sometimes large lists of parameters with no relevance to
+        the containing query being passed up. Rather than do this, we can
+        instead "render" the subqueries prior to using as a subquery, so our
+        containing code is only responsible for it's parameters, and we can
+        avoid any potential param collisions.
+        """
+        return cast(SyncClient, ch_client).substitute_params(query, params)
```

Comment:
Will need to pay attention if there are any moments where we rely on the overwriting of parameters by the final function to provide the right params.

Reply:
Yeah, I think if that's the case then there is a super strong dependency that we should make explicit, although not exactly sure what that should look like. Maybe just pulling up that param generation to the containing query.

🤞 the tests pick those cases up. I've purposefully avoided updating the non-breakdown query so hopefully even if not caught by a test this change will not cause a regression :famouslastwords: